src/middleware/companyAuth.middleware.ts

```diff
@@ -0,0 +1,94 @@
+import ApplicationLogger from 'ch-logging/lib/ApplicationLogger'
+import { Session } from 'ch-node-session-handler'
+import { CookieConfig } from 'ch-node-session-handler/lib/config/CookieConfig'
+import { SessionKey } from 'ch-node-session-handler/lib/session/keys/SessionKey'
+import { SignInInfoKeys } from 'ch-node-session-handler/lib/session/keys/SignInInfoKeys'
+import { ISignInInfo } from 'ch-node-session-handler/lib/session/model/SessionInterfaces'
+import { NextFunction, Request, RequestHandler, Response } from 'express'
+
+import { Mutable } from 'app/models/mutable'
+import { generateNonce, jweEncodeWithNonce } from 'app/utils/jwt.encryption'
+
+export interface AuthConfig {
+  accountUrl: string,
+  accountRequestKey: string,
+  accountClientId: string,
+  chsUrl: string,
+}
+
+interface ISignInInfoWithCompanyNumber extends ISignInInfo {
+  [SignInInfoKeys.CompanyNumber]?: string
+}
+
+const OATH_SCOPE_PREFIX = 'https://api.companieshouse.gov.uk/company/'
+
+export default function CompanyAuthMiddleware(config: CookieConfig, authConfig: AuthConfig,
+                                              logger: ApplicationLogger): RequestHandler {
+  return async (req: Request, res: Response, next: NextFunction) => {
+    const companyNumber = getCompanyNumberFromPath(req.originalUrl)
```

Comment:
It would be good to agree what format we will be retrieving the company number in:
1. Session
2. Query Params
3. Path params (I would avoid due to regex requirement)